### ISSUE TITLE
Critically missing linearization

### ISSUE THREAD

**Issue by virtualritz** [Traverse-Research/ispc-downsampler#25]
I briefly looked at the code and seems there is no linearization happening before downsampling. `u8`/channel color data will always have some gamma for obvious reasons. 

I.e. the results of the downsampling using this crate are just always wrong (as there are no `u16`/`f16` or `f32` images supported where there is enough bits to store data linearly in the pixel buffer).

See http://www.ericbrasseur.org/gamma.html for a detailed explanation.

This is the test image from above website:
![gamma_dalai_lama_gray](https://github.com/Traverse-Research/ispc-downsampler/assets/387036/47b6f022-5b7d-45ef-8967-bff7d4410ef3)

The expected result:
![gamma_dalai_lama_gray_good](https://github.com/Traverse-Research/ispc-downsampler/assets/387036/ff5ae266-1057-4d1b-a41b-cff1057c9410)

And what the code in this crate produces instead:
![gamma_dalai_lama_gray_result](https://github.com/Traverse-Research/ispc-downsampler/assets/387036/60a6de7c-ee98-4522-bbf4-31d4121bae32)

**Comment by Jasper-Bekkers** [comment]
I'll add a small note that this is true for colored images like albedo maps and the image above, but it may not be true for normal maps or roughness maps that don't contain perceptual color images (e.g. they're not in gamma space to begin with).

**Comment by virtualritz** [comment]
Indeed. It also only applies to downsampling. Upsampling gamma-corrected data should be just fine.
I.e. the note would be sth like:

> Do not use this to downsample images containing non-linear data, e.g. display color spaces like sRGB.

**Comment by MarijnS95** [comment]
@virtualritz Thank you for these resources: I'm not well-versed in image conversion algorithms but will give them a good read.

As expected, I think I have found the mistake that is happening in the code: we apply a 7x7 kernel to read source pixels and write them back to the target image, but this kernel is applied _to the target texture dimensions_ instead of the source texture dimensions. Meaning that when we scale (normalized, in float...) coordinates from the target to the source dimensions (between which is a 2:1 ratio), we only sample odd or even rows/columns, hence resulting in this color pattern!

https://github.com/Traverse-Research/ispc-downsampler/blob/d6747238e8799697365e258e7cd7d8532d551060/src/ispc/kernels/lanczos3.ispc#L71-L72

**Comment by MarijnS95** [comment]
https://github.com/Traverse-Research/ispc-downsampler/compare/lanczos3-kernel-range

Quickly cobbled together the two changes in minimal form to test this out, still a bunch TBD:

- Understand, cleanup and remove the various "center"s and float calculations;
- Stop clamping at the boundary edges, just skip out-of-bounds pixels?
- Provide parameters to optionally linearize the input and output (would be efficient to do in the ISPC kernel as opposed to requiring it to be linearized from Rust).

![square_test_result](https://github.com/Traverse-Research/ispc-downsampler/assets/2325264/c3570307-e81c-45aa-a941-1b0ff140c758)

**Comment by KYovchevski** [comment]
For the linearization: I already have a local branch that introduces new formats for gamma corrected textures, as well as a new kernel for the linearization which we can invoke depending on the format.

**Comment by virtualritz** [comment]
> Stop clamping at the boundary edges, just skip out-of-bounds pixels?

All texture resizing tools I ever used (and that's a ton) require a boundary mode to be specified during downsampling. Usually for mip-map generation.

The modes are usually (specified for u & v separately, if needed).
* Black (pixels outside the image window are zero)
* Clamp
* Periodic (wrap to the other side)

The reason is that later, when you look up the texture, the texture sampler needs to know what to do and if you have an override that contradicts what was done during mip level generation, you may get artifacts. 
This is often especially visible on displacement/bump/normal maps that are meant to tile (i.e. are  'periodic').

For downsampling normal images, e.g. photos you definitely want clamp though.

**Comment by MarijnS95** [comment]
Hi @virtualritz, thanks for the insights once again and apologies for forgetting about this issue. I've started solving all the other problems found in this crate and have now exposed a `gamma`/`degamma` parameter to the input and output of this crate to expose this behaviour:

https://github.com/Traverse-Research/ispc-downsampler/compare/degamma

Unfortunately there's still some color shifting, and I haven't investigated whether that might be caused by our implementation of the `lanczos3` filter and/or the `0.5` offset when we sample it. If I remove that offset, the current pixel is weighted `1` and the others `0`, resulting in a fully pink image because only even (or odd) pixels are taken into the final result:

https://www.desmos.com/calculator/rxasm8o0ne

![square_test_result](https://github.com/Traverse-Research/ispc-downsampler/assets/2325264/e5f247ac-feb5-4792-ba23-71c26ed575ec)

**Comment by virtualritz** [comment]
This already looks leaps and bounds better than before.

The only suggestion I have is to look at/compare with the [resp. filter code in OIIO](https://github.com/OpenImageIO/oiio/blob/a5c7a0c942ed439e53323b3315c868999ab29ec6/src/libutil/filter.cpp).
